# Phase 2: Differential Gene Expression Analysis
**Shreya Das**

In this phase we are going to show which genes are upregulated and downregulated in the different canacer groups. In our previous phase, we only figured out **which** genes are differentially expressed. In this phase we are more interested in **how** these genes are differentially expressed.

## Read-in the EDA Statistics .csv files

We are going to read-in the csv files from the previous phase into this notebook to use.

In [1]:
import pandas as pd
df_stats_ependymoma = pd.read_csv("EDA_stats_ependymoma")
df_stats_glioblastoma = pd.read_csv("EDA_stats_glioblastoma")
df_stats_medulloblastoma = pd.read_csv("EDA_stats_medulloblastoma")
df_stats_pilocytic_astrocytoma = pd.read_csv("EDA_stats_pilocytic_astrocytoma")

From here we are going to create a new Dataframe that only contains the genes that have statistically significant difference. Remember that this was indicated with the interpretation that we coded in the Welch's t-test in the previous phase.

In [2]:
# We define a function which takes a dataframe and then returns a dataframe which only have the rows with the interpretation
# of 'This gene is differentially expressed. (Fail to reject H0)'
def extract_stat_sig (daf):
    return pd.DataFrame(daf.loc[daf['Interpretation'] == 'This gene is differentially expressed. (Reject H0)'])


In [3]:
# We define a new dataframe and feed the output of the extract_stat_sig function into this defintiion
df_stats_sig_ependymoma = extract_stat_sig(df_stats_ependymoma)

# We check if the function worked
df_stats_sig_ependymoma

Unnamed: 0                  Gene  Statistics             p  alpha  \
0               0             1007_s_at    8.549862  4.551427e-07   0.05   
1               1               1053_at    3.010169  7.539762e-03   0.05   
2               2                117_at    2.946339  7.094482e-03   0.05   
4               4             1255_g_at   -5.365298  9.955036e-05   0.05   
5               5               1294_at    5.547675  3.892620e-06   0.05   
...           ...                   ...         ...           ...    ...   
54664       54664  AFFX-r2-Ec-bioC-5_at   -4.064777  4.474298e-04   0.05   
54665       54665  AFFX-r2-Ec-bioD-3_at   -3.720401  7.438247e-04   0.05   
54666       54666  AFFX-r2-Ec-bioD-5_at   -3.292114  3.349294e-03   0.05   
54667       54667   AFFX-r2-P1-cre-3_at   -3.400711  2.880842e-03   0.05   
54668       54668   AFFX-r2-P1-cre-5_at   -3.340604  3.511498e-03   0.05   

                                          Interpretation  
0      This gene is differentially expressed. (Reject...  
1      This gene is differentially expressed. (Reject...  
2      This gene is differentially expressed. (Reject...  
4      This gene is differentially expressed. (Reject...  
5      This gene is differentially expressed. (Reject...  
...                                                  ...  
54664  This gene is differentially expressed. (Reject...  
54665  This gene is differentially expressed. (Reject...  
54666  This gene is differentially expressed. (Reject...  
54667  This gene is differentially expressed. (Reject...  
54668  This gene is differentially expressed. (Reject...  

[27039 rows x 6 columns]

We see that for the ependymoma group, there is 27039 genes are found to be statistically significant compared to the control group.

Let's do the same for the other 3 cancer groups.

## Glioblastoma

In [4]:
df_stats_sig_glioblastoma = extract_stat_sig(df_stats_glioblastoma)

df_stats_sig_glioblastoma

Unnamed: 0                  Gene  Statistics         p  alpha  \
0               0             1007_s_at    6.109568  0.000008   0.05   
1               1               1053_at    6.284144  0.000002   0.05   
2               2                117_at    3.759057  0.001005   0.05   
4               4             1255_g_at   -3.576847  0.002181   0.05   
5               5               1294_at    4.254927  0.000124   0.05   
...           ...                   ...         ...       ...    ...   
54664       54664  AFFX-r2-Ec-bioC-5_at   -3.861179  0.000924   0.05   
54665       54665  AFFX-r2-Ec-bioD-3_at   -3.495233  0.002357   0.05   
54666       54666  AFFX-r2-Ec-bioD-5_at   -2.871692  0.010224   0.05   
54667       54667   AFFX-r2-P1-cre-3_at   -4.105825  0.000645   0.05   
54668       54668   AFFX-r2-P1-cre-5_at   -4.393207  0.000314   0.05   

                                          Interpretation  
0      This gene is differentially expressed. (Reject...  
1      This gene is differentially expressed. (Reject...  
2      This gene is differentially expressed. (Reject...  
4      This gene is differentially expressed. (Reject...  
5      This gene is differentially expressed. (Reject...  
...                                                  ...  
54664  This gene is differentially expressed. (Reject...  
54665  This gene is differentially expressed. (Reject...  
54666  This gene is differentially expressed. (Reject...  
54667  This gene is differentially expressed. (Reject...  
54668  This gene is differentially expressed. (Reject...  

[24959 rows x 6 columns]

## Medulloblastoma

In [5]:
df_stats_sig_medulloblastoma = extract_stat_sig(df_stats_medulloblastoma)

df_stats_sig_medulloblastoma

Unnamed: 0                  Gene  Statistics         p  alpha  \
1               1               1053_at    5.422061  0.000016   0.05   
4               4             1255_g_at   -4.627142  0.000253   0.05   
6               6               1316_at   -2.587390  0.015684   0.05   
7               7               1320_at    3.436782  0.001703   0.05   
9               9               1431_at   -4.638950  0.000145   0.05   
...           ...                   ...         ...       ...    ...   
54647       54647        AFFX-PheX-M_at   -2.882278  0.010038   0.05   
54656       54656   AFFX-r2-Bs-phe-M_at   -3.203696  0.004757   0.05   
54660       54660  AFFX-r2-Ec-bioB-3_at   -2.100233  0.044043   0.05   
54667       54667   AFFX-r2-P1-cre-3_at   -2.751830  0.011081   0.05   
54668       54668   AFFX-r2-P1-cre-5_at   -2.903066  0.007426   0.05   

                                          Interpretation  
1      This gene is differentially expressed. (Reject...  
4      This gene is differentially expressed. (Reject...  
6      This gene is differentially expressed. (Reject...  
7      This gene is differentially expressed. (Reject...  
9      This gene is differentially expressed. (Reject...  
...                                                  ...  
54647  This gene is differentially expressed. (Reject...  
54656  This gene is differentially expressed. (Reject...  
54660  This gene is differentially expressed. (Reject...  
54667  This gene is differentially expressed. (Reject...  
54668  This gene is differentially expressed. (Reject...  

[23631 rows x 6 columns]

## Pilocytic Astrocytoma

In [6]:
df_stats_sig_pilo_astrocytoma = extract_stat_sig(df_stats_pilocytic_astrocytoma)

df_stats_sig_pilo_astrocytoma

Unnamed: 0                  Gene  Statistics             p  alpha  \
0               0             1007_s_at    9.059796  6.470481e-08   0.05   
5               5               1294_at    6.806438  3.179949e-07   0.05   
6               6               1316_at   -3.000466  6.168611e-03   0.05   
7               7               1320_at    3.107922  4.691813e-03   0.05   
8               8             1405_i_at    2.885509  7.835963e-03   0.05   
...           ...                   ...         ...           ...    ...   
54661       54661  AFFX-r2-Ec-bioB-5_at   -2.333800  2.773792e-02   0.05   
54663       54663  AFFX-r2-Ec-bioC-3_at   -2.345115  2.725426e-02   0.05   
54664       54664  AFFX-r2-Ec-bioC-5_at   -2.211730  3.601706e-02   0.05   
54667       54667   AFFX-r2-P1-cre-3_at   -2.076973  4.805526e-02   0.05   
54668       54668   AFFX-r2-P1-cre-5_at   -2.173326  3.913387e-02   0.05   

                                          Interpretation  
0      This gene is differentially expressed. (Reject...  
5      This gene is differentially expressed. (Reject...  
6      This gene is differentially expressed. (Reject...  
7      This gene is differentially expressed. (Reject...  
8      This gene is differentially expressed. (Reject...  
...                                                  ...  
54661  This gene is differentially expressed. (Reject...  
54663  This gene is differentially expressed. (Reject...  
54664  This gene is differentially expressed. (Reject...  
54667  This gene is differentially expressed. (Reject...  
54668  This gene is differentially expressed. (Reject...  

[22460 rows x 6 columns]

We can check the number by checking the countplots that we generated in the previous phase, just to make sure that we choose the right data to select.

## Calculation of Increased or Decreased Gene Expression

In order to understand which statistically significant genes have increased or decreased gene expression, we need to analyze the t-test that we calculated in the previous phase. 

Generally, we would calculate fold-change for the gene expression. However, in our current dataset the values that are listed under each genes are not counts of RNA transcripts for each of the genes and are rather gene expression data from multiple sources of DNA microarrays. From the published paper, the authors stated that very high values of gene expression were interpreted as 'high gene expression' and negative or low values of gene expression were interpreted as 'low gene expression'. Quickly looking at the original data, we see that there are no negative values.

Looking at the statistics that we calculated, the t-test has both negative and positive values. The sign of the t-test statistic is not relevant to the level of significance (in most cases, people will only take the absolute value of the t-statistic), but rather defines which group has the lower mean (if the value is negative, this means that the first group has a lower mean than the second group). In the Wlech's t-test function that we program, the first group given is the experimental group (cancer) and the second group is the control group (healthy). This means we can interprete a negative Welch's t-test value as a reversal of the directionalty of the effect (gene expression) being studied, or that gene expression is lower in the experimental group compared to the control group. 

In [7]:
# First let's grab the statistics only from these 4 dataframes with the associated gene names
selection = ['Gene', 'Statistics']

# Ependymoma
df_expression_ependymoma = df_stats_sig_ependymoma[selection]

# Glioblastoma
df_expression_glioblastoma = df_stats_sig_glioblastoma[selection]

# Medulloblastoma
df_expression_medulloblastoma = df_stats_sig_medulloblastoma[selection]

# Pilocytic Astrocytoma
df_expression_pilo_astro = df_stats_sig_pilo_astrocytoma[selection]


We can check that the selection worked:

In [8]:
df_expression_medulloblastoma

Gene  Statistics
1                   1053_at    5.422061
4                 1255_g_at   -4.627142
6                   1316_at   -2.587390
7                   1320_at    3.436782
9                   1431_at   -4.638950
...                     ...         ...
54647        AFFX-PheX-M_at   -2.882278
54656   AFFX-r2-Bs-phe-M_at   -3.203696
54660  AFFX-r2-Ec-bioB-3_at   -2.100233
54667   AFFX-r2-P1-cre-3_at   -2.751830
54668   AFFX-r2-P1-cre-5_at   -2.903066

[23631 rows x 2 columns]

We are just going to rename the statistics column with the cancer group. Let's do it for the other dataframes too.

In [67]:
df_expression_ependymoma.rename(columns = {'Statistics': 'Ependymoma'}, inplace = True)

df_expression_medulloblastoma.rename(columns = {'Statistics': 'Medulloblastoma'}, inplace = True)

df_expression_glioblastoma.rename(columns = {'Statistics': 'Glioblastoma'}, inplace = True)

df_expression_pilo_astro.rename(columns = {'Statistics': 'Pilocytic Astrocytoma'}, inplace = True)

/var/folders/6r/nx6jkx254q31jrf83vmvvwhw0000gn/T/ipykernel_1796/2381849600.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_expression_ependymoma.rename(columns = {'Statistics': 'Ependymoma'}, inplace = True)
/var/folders/6r/nx6jkx254q31jrf83vmvvwhw0000gn/T/ipykernel_1796/2381849600.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_expression_medulloblastoma.rename(columns = {'Statistics': 'Medulloblastoma'}, inplace = True)
/var/folders/6r/nx6jkx254q31jrf83vmvvwhw0000gn/T/ipykernel_1796/2381849600.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the cavea

In [68]:
genes = pd.read_csv('df_stats_ependymoma')

gene_ID = genes['Gene']

gene_ID

0              1007_s_at
1                1053_at
2                 117_at
3                 121_at
4              1255_g_at
              ...       
54670     AFFX-ThrX-5_at
54671     AFFX-ThrX-M_at
54672    AFFX-TrpnX-3_at
54673    AFFX-TrpnX-5_at
54674    AFFX-TrpnX-M_at
Name: Gene, Length: 54675, dtype: object

In [69]:
df = pd.concat([gene_ID, df_expression_ependymoma, df_expression_glioblastoma, df_expression_medulloblastoma, df_expression_pilo_astro], axis = 1)
df

Gene       Gene  Ependymoma       Gene  Glioblastoma  \
0            1007_s_at  1007_s_at    8.549862  1007_s_at      6.109568   
1              1053_at    1053_at    3.010169    1053_at      6.284144   
2               117_at     117_at    2.946339     117_at      3.759057   
3               121_at        NaN         NaN        NaN           NaN   
4            1255_g_at  1255_g_at   -5.365298  1255_g_at     -3.576847   
...                ...        ...         ...        ...           ...   
54670   AFFX-ThrX-5_at        NaN         NaN        NaN           NaN   
54671   AFFX-ThrX-M_at        NaN         NaN        NaN           NaN   
54672  AFFX-TrpnX-3_at        NaN         NaN        NaN           NaN   
54673  AFFX-TrpnX-5_at        NaN         NaN        NaN           NaN   
54674  AFFX-TrpnX-M_at        NaN         NaN        NaN           NaN   

            Gene  Medulloblastoma       Gene  Pilocytic Astrocytoma  
0            NaN              NaN  1007_s_at               9.059796  
1        1053_at         5.422061        NaN                    NaN  
2            NaN              NaN        NaN                    NaN  
3            NaN              NaN        NaN                    NaN  
4      1255_g_at        -4.627142        NaN                    NaN  
...          ...              ...        ...                    ...  
54670        NaN              NaN        NaN                    NaN  
54671        NaN              NaN        NaN                    NaN  
54672        NaN              NaN        NaN                    NaN  
54673        NaN              NaN        NaN                    NaN  
54674        NaN              NaN        NaN                    NaN  

[54675 rows x 9 columns]

In [70]:
df = df.iloc[:,[0, 2, 4, 6, 8]]

df

Gene  Ependymoma  Glioblastoma  Medulloblastoma  \
0            1007_s_at    8.549862      6.109568              NaN   
1              1053_at    3.010169      6.284144         5.422061   
2               117_at    2.946339      3.759057              NaN   
3               121_at         NaN           NaN              NaN   
4            1255_g_at   -5.365298     -3.576847        -4.627142   
...                ...         ...           ...              ...   
54670   AFFX-ThrX-5_at         NaN           NaN              NaN   
54671   AFFX-ThrX-M_at         NaN           NaN              NaN   
54672  AFFX-TrpnX-3_at         NaN           NaN              NaN   
54673  AFFX-TrpnX-5_at         NaN           NaN              NaN   
54674  AFFX-TrpnX-M_at         NaN           NaN              NaN   

       Pilocytic Astrocytoma  
0                   9.059796  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  
...                      ...  
54670                    NaN  
54671                    NaN  
54672                    NaN  
54673                    NaN  
54674                    NaN  

[54675 rows x 5 columns]

In [71]:
df = df.fillna(0)

df

Gene  Ependymoma  Glioblastoma  Medulloblastoma  \
0            1007_s_at    8.549862      6.109568         0.000000   
1              1053_at    3.010169      6.284144         5.422061   
2               117_at    2.946339      3.759057         0.000000   
3               121_at    0.000000      0.000000         0.000000   
4            1255_g_at   -5.365298     -3.576847        -4.627142   
...                ...         ...           ...              ...   
54670   AFFX-ThrX-5_at    0.000000      0.000000         0.000000   
54671   AFFX-ThrX-M_at    0.000000      0.000000         0.000000   
54672  AFFX-TrpnX-3_at    0.000000      0.000000         0.000000   
54673  AFFX-TrpnX-5_at    0.000000      0.000000         0.000000   
54674  AFFX-TrpnX-M_at    0.000000      0.000000         0.000000   

       Pilocytic Astrocytoma  
0                   9.059796  
1                   0.000000  
2                   0.000000  
3                   0.000000  
4                   0.000000  
...                      ...  
54670               0.000000  
54671               0.000000  
54672               0.000000  
54673               0.000000  
54674               0.000000  

[54675 rows x 5 columns]

## Generating Heat Maps
In order to visualize the differential gene expression, we will use a heat map. Heat maps are generally used to visualize which genes are highly expressed and which ones are lowly expressed.

Through this method which also figure out the top 5 genes that are highly expressed in each cancer group just by looking at the color.

To do this, I will be using R to generate the heat map. We need to export the dataframe that we generated here.

In [79]:
df.to_csv('Gene Expression Stats.csv')